diff --git a/util-linux/lspci.c b/util-linux/lspci.c
index a6b4e85..45e741d 100644
--- a/util-linux/lspci.c
+++ b/util-linux/lspci.c
@@ -19,6 +19,7 @@
 //applet:IF_LSPCI(APPLET_NOEXEC(lspci, lspci, BB_DIR_USR_BIN, BB_SUID_DROP, lspci))
 
 //kbuild:lib-$(CONFIG_LSPCI) += lspci.o
+//kbuild:lib-$(CONFIG_LSPCI) += sprintid.o
 
 //usage:#define lspci_trivial_usage
 //usage:       "[-mk]"
@@ -35,6 +36,8 @@ enum {
 	OPT_k = (1 << 1),
 };
 
+#include "sprintid.h"
+
 /*
  * PCI_SLOT_NAME PCI_CLASS: PCI_VID:PCI_DID [PCI_SUBSYS_VID:PCI_SUBSYS_DID] [DRIVER]
  */
@@ -49,6 +52,7 @@ static int FAST_FUNC fileAction(
 	char *pci_slot_name = NULL, *driver = NULL;
 	int pci_class = 0, pci_vid = 0, pci_did = 0;
 	int pci_subsys_vid = 0, pci_subsys_did = 0;
+	char *pci_human_readable;
 
 	char *uevent_filename = concat_path_file(fileName, "/uevent");
 	parser = config_open2(uevent_filename, fopen_for_read);
@@ -90,8 +94,15 @@ static int FAST_FUNC fileAction(
 			pci_slot_name, pci_class, pci_vid, pci_did,
 			pci_subsys_vid, pci_subsys_did);
 	} else {
+		pci_human_readable = sprintid(pci_class, pci_vid, pci_did);
+		if (pci_human_readable != NULL) {
+			printf("%s %s", pci_slot_name, pci_human_readable);
+		}
+		else
+		{
 		printf("%s Class %04x: %04x:%04x",
 			pci_slot_name, pci_class, pci_vid, pci_did);
+		}
 	}
 
 	if ((option_mask32 & OPT_k) && driver) {
@@ -114,6 +125,7 @@ int lspci_main(int argc UNUSED_PARAM, char **argv)
 {
 	getopt32(argv, "m" /*non-compat:*/ "k" /*ignored:*/ "nv");
 
+	sprintid_init("/usr/share/hwdata/pci.ids");
 	recursive_action("/sys/bus/pci/devices",
 			ACTION_RECURSE,
 			fileAction,
@@ -121,5 +133,6 @@ int lspci_main(int argc UNUSED_PARAM, char **argv)
 			NULL, /* userData */
 			0 /* depth */);
 
+	sprintid_cleanup();
 	return EXIT_SUCCESS;
 }
diff --git a/util-linux/lsusb.c b/util-linux/lsusb.c
index e27aa7f..3a16fec 100644
--- a/util-linux/lsusb.c
+++ b/util-linux/lsusb.c
@@ -19,11 +19,13 @@
 //applet:IF_LSUSB(APPLET_NOEXEC(lsusb, lsusb, BB_DIR_USR_BIN, BB_SUID_DROP, lsusb))
 
 //kbuild:lib-$(CONFIG_LSUSB) += lsusb.o
+//kbuild:lib-$(CONFIG_LSUSB) += sprintid.o
 
 //usage:#define lsusb_trivial_usage NOUSAGE_STR
 //usage:#define lsusb_full_usage ""
 
 #include "libbb.h"
+#include "sprintid.h"
 
 static int FAST_FUNC fileAction(
 		const char *fileName,
@@ -35,6 +37,7 @@ static int FAST_FUNC fileAction(
 	char *tokens[4];
 	char *busnum = NULL, *devnum = NULL;
 	int product_vid = 0, product_did = 0;
+	char *usb_human_readable;
 	char *uevent_filename = concat_path_file(fileName, "/uevent");
 
 	parser = config_open2(uevent_filename, fopen_for_read);
@@ -64,7 +67,12 @@ static int FAST_FUNC fileAction(
 	config_close(parser);
 
 	if (busnum) {
+		usb_human_readable = sprintid(-1, product_vid, product_did);
+		if (usb_human_readable != NULL) {
+			printf("Bus %s Device %s: ID %04x:%04x %s\n", busnum, devnum, product_vid, product_did, usb_human_readable);
+		} else {
 		printf("Bus %s Device %s: ID %04x:%04x\n", busnum, devnum, product_vid, product_did);
+		}
 		free(busnum);
 		free(devnum);
 	}
@@ -77,6 +85,7 @@ int lsusb_main(int argc UNUSED_PARAM, char **argv UNUSED_PARAM)
 {
 	/* no options, no getopt */
 
+	sprintid_init("/usr/share/hwdata/usb.ids");
 	recursive_action("/sys/bus/usb/devices",
 			ACTION_RECURSE,
 			fileAction,
@@ -84,5 +93,6 @@ int lsusb_main(int argc UNUSED_PARAM, char **argv UNUSED_PARAM)
 			NULL, /* userData */
 			0 /* depth */);
 
+	sprintid_cleanup();
 	return EXIT_SUCCESS;
 }
diff --git a/util-linux/sprintid.c b/util-linux/sprintid.c
new file mode 100644
index 0000000..bb8a26b
--- /dev/null
+++ b/util-linux/sprintid.c
@@ -0,0 +1,110 @@
+#include "sprintid.h"
+
+static FILE *sprintid_db;
+static char *sprintid_line;
+
+#define SPRINTID_NAMELEN 128
+
+inline int sprintid_init(char *dbname) {
+   sprintid_db=fopen(dbname,"r");
+   sprintid_line=malloc(3*SPRINTID_NAMELEN);
+   if(sprintid_db==NULL || sprintid_line==NULL)
+       return(1);
+   else
+       return(0);
+}
+
+inline int sprintid_cleanup(void) {
+   if(sprintid_db!=NULL) {
+       fclose(sprintid_db);
+   }
+   if(sprintid_line!=NULL) {
+       free(sprintid_line);
+   }
+   return(0);
+}
+
+char *sprintid(int cidnum, int vidnum, int didnum) {
+   char line[3*SPRINTID_NAMELEN+1];
+   char vendor[SPRINTID_NAMELEN+1];
+   char device[SPRINTID_NAMELEN+1];
+   char class[SPRINTID_NAMELEN+1];
+
+   char cid[5];
+   char vid[5];
+   char did[5];
+
+   int havedevice=0;
+
+   if(sprintid_line==NULL) {
+       return(NULL);
+   }
+
+   sprintf(vid,"%04x",vidnum);
+   sprintf(did,"%04x",didnum);
+
+   if(cidnum>0)
+   {
+       sprintf(cid,"%04x",cidnum);
+       sprintf(sprintid_line,"Class %s: %s:%s",cid,vid,did);
+   } else {
+       sprintf(cid,"");
+       sprintf(sprintid_line,"%s:%s",vid,did);
+   }
+
+   if(sprintid_db==NULL) {
+       return(sprintid_line);
+   }
+
+   if(cidnum>0) {
+       sprintf(class,"Class unknown class %s: ",cid);
+   }
+   sprintf(vendor,"unknown vendor %s",vid);
+   sprintf(device,"unknown device %s",did);
+
+   fseek(sprintid_db,0L,SEEK_SET);
+
+   while(fgets(line, 3*SPRINTID_NAMELEN, sprintid_db)!=NULL) {
+       if(line[0]!='#' && line[0]!='\r' && line[0]!='\n')
+       {
+               if((havedevice==0) && ((line[0]>='0' && line[0]<='9') || (line[0]>='a' && line[0]<='f'))) { // Vendor
+               if(strncmp(line,vid,4)==0) {
+                   strncpy(vendor,&line[6],SPRINTID_NAMELEN);
+                   while(fgets(line, 3*SPRINTID_NAMELEN, sprintid_db)!=NULL && !((line[0]>='0' && line[0]<='9') || (line[0]>='a' && line[0]<='f'))) {
+                       if(line[0]=='\t' && ((line[1]>='0' && line[1]<='9') || (line[1]>='a' && line[1]<='f'))) { // Device
+                           if(strncmp(&line[1],did,4)==0) {
+                               strncpy(device,&line[7],SPRINTID_NAMELEN);
+                               break;
+                           }
+                       }
+                   }
+               }
+           }
+
+           if(cidnum>0 && line[0]=='C') { // Classes
+               if(strncmp(&line[2],cid,2)==0) { // found Major
+                   while(fgets(line, 3*SPRINTID_NAMELEN, sprintid_db)!=NULL && !(line[0]=='C')) {
+                       if(line[0]=='\t' && ((line[1]>='0' && line[1]<='9') || (line[1]>='a' && line[1]<='f'))) { // subclass
+                           if(strncmp(&line[1],&cid[2],2)==0) {
+                               strncpy(class,&line[5],SPRINTID_NAMELEN);
+                               break;
+                           }
+                       }
+                   }
+                   break;
+               }
+           }
+       }
+   }
+   chomp(class);
+   chomp(vendor);
+   chomp(device);
+   if(cidnum>0)
+   {
+       sprintf(sprintid_line,"Class %s: %s:%s",class,vendor,device);
+   } else {
+       sprintf(sprintid_line,"%s:%s",vendor,device);
+   }
+   return(sprintid_line);
+}
+
diff --git a/util-linux/sprintid.h b/util-linux/sprintid.h
new file mode 100644
index 0000000..5a9e309
--- /dev/null
+++ b/util-linux/sprintid.h
@@ -0,0 +1,7 @@
+#include <stdio.h>
+#include <string.h>
+#include <libbb.h>
+
+extern char *sprintid(int cidnum, int vidnum, int didnum);
+extern int sprintid_init(char *dbname);
+extern int sprintid_cleanup(void);
